Better user-key documentation directions

FontouraAbreu opened the pull request:
### Description

The following files where modified:

- README.md
- docs/models/userkey.md

The following files where created:

- assets/first-user-key.png

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [ ] Other (please describe):

### Test Cases

- [ ] Test cases are covered for this PR.
- [x] This PR does not require test cases (please explain): Documentation update only

### Additional Information

I made the changes based on a difficulty I had when first-time configuring secrets. I did find the proper documentation by `docs/models/userkey.md` but nothing was pointing to it at the `README.md` and nothing was said about the need to click on the save button for properly saving my new generated keys.

abhi1693 commented on README.md:
```diff
@@ -36,54 +47,61 @@ and stability.
 
+## User Keys
+
+Each user within NetBox can associate his or her account with an RSA public key. If activated by an administrator, this user key will contain a unique, encrypted copy of the AES master key needed to retrieve secret data.
```
```suggestion
Each user within NetBox can associate their account with an RSA public key. If activated by an administrator, this user key will contain a unique, encrypted copy of the AES master key needed to retrieve secret data.
```

I think it's also better to mention activation is one time

abhi1693 commented on README.md:
```diff
@@ -36,54 +47,61 @@ and stability.
+
+Each user within NetBox can associate his or her account with an RSA public key. If activated by an administrator, this user key will contain a unique, encrypted copy of the AES master key needed to retrieve secret data.
+
+User keys may be created by users individually, however they are of no use until they have been activated by a user who already possesses an active user key.
```
This sounds a bit redundant to me. It's better to rephrase it

FontouraAbreu replied:
This was copied from `docs/models/userkey.md`. I personally Prefer every usage or deployment related inside the main `Readme.md` file and only development documentation inside the `docs` directory.

I will bring everything into the `Readme.md` file and you tell me what you think of it

abhi1693 reviewed (changes requested):
For now, I have some minor nit picks.

FontouraAbreu commented:
Sure thing!

Will make the modifications and also add some more usage examples using pynetbox.

FontouraAbreu commented:
I've made the changes as well as added a whole `Pynetbox` examples section.
Every documentation was merged into the main `Readme.md` file.

I fixed an error where the API endpoints where `<netbox_url>/api/plugins/secrets/secrets/secrets/<any>` instead of `<netbox_url>/api/plugins/secrets/secrets/<any>`. Also changed an incorrect "Netbox Admin UI" path that was removed from netbox in a previous update to a sidebar menu.

I changed some text in order to make the documentation a bit more readable.

Please check it out, specially the Pynetbox section. 

Obs.: Not every pynetbox example has been tested, in order to do these i'd like first your opinion on the documentation arrangement :)

an-ky commented on README.md:
```diff
@@ -36,54 +66,624 @@ and stability.
+    "plaintext": "foobar",
+}
+
+nb.plugins.secrets.secrets.create(data=secret, session=session)
```
Sorry for stepping in but I found an error calling the pynetbox API here:

Using this line from the example, I got 
```
ynetbox.core.query.RequestError: The request failed with code 400 Bad Request: {'assigned_object_type': ['This field is required.'], 'assigned_object_id': ['This field is required.'], 'role': ['This field is required.'], 'plaintext': ['This field is required.']}
```
After trying around a bit, I got it working by using this call instead `nb.plugins.secrets.secrets.create(secret)` because create() is defined as `create(*args, **kwargs)`, taking the properties of the objects to be created directly, or a list of them as dicts. See https://pynetbox.readthedocs.io/en/latest/endpoint.html#pynetbox.core.endpoint.Endpoint.create

Nevertheless this documentation enhancement is awesome.

FontouraAbreu replied:
Good one!

I've not tested everything, just put on what I thought would work. I'll take some time to test everything up and update the pull request.

Thanks for the correction!

abhi1693 commented:
Before releasing the plugin for v4.2, I'll go through this doc and make sure it's good to merge. I'd rather have this change with the newer release than add to existing ones.

FontouraAbreu commented:
That's perfect!

I'm aiming to upgrade to v4.2 and `netbox-secrets` is the only plugin left to enable the upgrade.

kprince28 commented on README.md:
````diff
@@ -36,54 +66,624 @@ and stability.
+
+```json
+{
+    "private_key": "<private key>"
````
I noticed that it doesn't mention the optional `preserve_key` field in the payload. This field allows users to reuse an existing session key instead of creating a new one.

kprince28 commented on README.md:
```diff
@@ -36,54 +66,624 @@ and stability.
+    "plaintext": "foobar",
+}
+
+nb.plugins.secrets.secrets.update(id=1, data=secret, session=session)
```
The `update` method requires the `id` to be included in the payload rather than as a parameter. Here is the correct payload example:
```
secret = {
    "id": 1,
    "assigned_object_type": "dcim.device",
    "assigned_object_id": 35,
    "role": 1,
    "name": "admin",
    "plaintext": "foobar",
}
nb.plugins.secrets.secrets.update([secret])
```
Make sure to pass the payload in this format when updating an instance.

If we want to update a single instance, we first need to fetch the instance and then update it like this:
```
x = nb.dcim.devices.get(1)
x.update({
    "name": "test-switch2",
    "serial": "ABC321",
})
```

kprince28 reviewed (changes requested):
Could you please thoroughly test the API before writing the documentation? I've noticed some mistakes in the API docs, and proper testing would help ensure accuracy and clarity.

FontouraAbreu commented on README.md:
````diff
@@ -36,54 +66,624 @@ and stability.
+
+```json
+{
+    "private_key": "<private key>"
````
okay

FontouraAbreu commented:
sure thing. I will aim to add a `tests` directory in order to enable unit tests for these API calls.

FontouraAbreu commented:
Yes I am, just had to focus on some other things at work. I'm hoping to test it in the following weeks.
